### DISCUSSION TITLE
BLE iOS best practice

### DISCUSSION THREAD

**Discussion by wesoos** [shinyorg/shiny#967] [Q&A] [answered]
Is there a best practice guideline for BLE on iOS devices?  I'm developing an app to work with a relay device, using Xamarin.Forms for Android and iOS.    Everything works fine with Android, I connect to the peripheral, write some data to pop the relay, receive a notification, write more data, receive a notification, disconnect from peripheral.  All this happens on a single button click.  I can do this over and over and over and works fine every time.

On iOS however, it only works the first time, then is starts flaking.  The second try, it will connect and write the first set of data, then after that it doesn't write anymore.  Connect and disconnect still works.  I've tried the subscribers, async/await commands, and ran everything on main thread, but same results.

Any advice or anything I need to look out for with iOS?

Thanks

**Comment by aritchie** [comment]
You don't need to run anything on the main thread.  That isn't documented anywhere that you need to do that.  As far as a best practice guide, no... I don't commit time into things like that.

As for the rest of your question, I'm not sure what to tell you as I don't have anywhere near enough info to even make a guess.

**Comment by wesoos** [comment]
Thanks. I was just wondering if you knew of something to look out for, and I only mentioned the main thread as something I tried during troubleshooting...

Regards

**Comment by wesoos** [comment, marked as answer]
I have the following code.  In Android this works fine, but on iOS, the notifications from the peripheral stops coming back after the first interaction.

This code is in my platform project.  From my standard project, I first call the "ConnectDevice", which after successful connection, it finds the correct service, then finds the correct characteristic, and then hooks in to receive notifications on that characteristic.

Then I write the first value to that characteristic ("A10101A1AA").  On success based on a notification received, I immediately write new data to the characteristic ("A10401A4AA").  Again on success based on the notification received, I disconnect from the peripheral.

All this happens on one button click.

```
public void ConnectDevice(BTDevice device)
        {
            currentNotificationResponse = "";
            timerNotifyCount = 0;

            device.Peripheral.WhenConnected().Subscribe(
                result =>
                {
                    device.Connected = result.IsConnected();

                    if (device.Connected)
                    {
                        currentConnectedDevice = device;

                        result.GetServices().Subscribe(
                            results =>
                            {
                                foreach (var service in results)
                                {
                                    if (service.Uuid.ToLower().Equals(UUID_SERVICE_STR.ToLower()))
                                    {
                                        service.GetCharacteristics().Subscribe(
                                            results =>
                                            {
                                                foreach (var characteristic in results)
                                                {
                                                    if (characteristic.Uuid.ToLower().Equals(UUID_CHAR_STR.ToLower()))
                                                    {
                                                        connectDeviceCharacteristic = characteristic;
                                                        connectDeviceCharacteristic.EnableNotifications(true, true).Subscribe(result => { }, exception => { });
                                                        connectDeviceCharacteristic.WhenNotificationReceived().Subscribe(NotificationReceived);

                                                        GlobalEventHandlers.OnBleDeviceConnected(new CustomEventArgs(device, null));

                                                        break;
                                                    }
                                                }
                                            },
                                            exception =>
                                            {

                                            });

                                        break;
                                    }
                                }

                                if (connectDeviceCharacteristic == null)
                                    DisconnectDevice();
                            },
                            exception =>
                            {

                            });
                    }
                });

            device.Peripheral.Connect(new ConnectionConfig { AutoConnect = false });

            //return true;
        }
```

```
private void NotificationReceived(GattCharacteristicResult result)
        {
            if (result.Data != null)
            {
                currentNotificationResponse += BitConverter.ToString(result.Data).Replace("-", "");
            }

            if (currentNotificationResponse.Equals("A10101A1AA")) //Relay turned on
            {
                currentNotificationResponse = "";

                //WriteValue(currentConnectedDevice, "A104D2040100000371AA", BleWriteCommand.CloseTimeRelay);

                GlobalEventHandlers.OnBleNotification(new BleWriteEventArgs(currentConnectedDevice.Address, BleWriteCommand.OpenRelay));
            }
            else if(currentNotificationResponse.Equals("A10201A2AA")) //Relay turned off
            {
                currentNotificationResponse = "";
                DisconnectDevice();
            }
            else if (currentNotificationResponse.Equals("A10401A4AA")) //Timer set to turn relay off
            {
                currentNotificationResponse = "";
                timerNotifyCount++;

                if(timerNotifyCount == 2)
                {
                    DisconnectDevice();
                }
            }
        }
```

```
public void DisconnectDevice()
        {
            currentNotificationResponse = "";
            timerNotifyCount = 0;

            if (currentConnectedDevice != null)
            {
                currentConnectedDevice.Peripheral.WhenDisconnected().Subscribe(
                    result =>
                    {
                        currentConnectedDevice.Connected = result.IsConnected();

                        if (!currentConnectedDevice.Connected)
                        {
                            currentConnectedDevice = null;
                            connectDeviceCharacteristic = null;

                            GlobalEventHandlers.OnBleDeviceDisConnected(new CustomEventArgs(null, null));
                        }
                    });

                currentConnectedDevice.Peripheral.CancelConnection();
            }
        }
```

**Reply by aritchie** [reply]
Nothing jumps out immediately as weird.  Where is it stopping?  Are there errors?  Is there anything in the console?

Generally iOS just works and Android is the problem platform

**Reply by wesoos** [reply]
No errors, nothing in the console.

Essentially this is the sequence upon button press:

- Connect to peripheral
- Find characteristic and hook to notifications
- Write first value
- Receive notification
- Write second value
- Receive notification
- Disconnect

The second write creates a timer on the peripheral of about 6 seconds.  Then it sends another notification when the timer expires.  After I receive this second notification of the second write, I disconnect.

For context, this is a relay device.  So the first write pops the relay.  The second write sets the device to close the relay after 6 seconds.  Then when the device closes the relay, is when it sends that second notification and I disconnect.

On iOS, it stops sending notifications after a second button press.  Meaning, I press the button the first time, it goes through the whole sequence fine.  Then I press the button again, and it does connect and it does the first write, but no more notifications.

When I kill the app and load again, it works again as above...

**Reply by aritchie** [reply]
This is the last of my guesses:
1) I only see one write, you mentioned two?
2) The subscription might be getting garbage collected.  This happens with deep nested observables sometimes.

I do TX/RX flows all the time without issue all the time.  I'd suggest taking a look at a sample where I do something similar here:  https://github.com/shinyorg/samples/blob/main/BluetoothLE-Managed/Sample/ManagedPeripheralViewModel.cs

Hopefully that helps.

**Reply by wesoos** [reply]
Hi Allan,

I used your managed sample code, and getting better consistency in terms of writing to peripheral now.  However I'm never getting any notifications period, iOS or Android.  If you get a chance, please see code below:

```
Peripheral = _SelectedDevice.Peripheral.CreateManaged();

                var sub = Peripheral
                        .WhenNotificationReceived(UUID_SERVICE_STR, UUID_CHAR_STR)
                        .SubOnMainThread(data =>
                        {
                            var txt = Encoding.ASCII.GetString(data);
                        });

                await Peripheral.EnableNotifications(
                    true,
                    UUID_SERVICE_STR,
                    UUID_CHAR_STR
                );

                await WriteValue("A104D2010177AA");
                await Task.Delay(1000);
                await WriteValue(GetTimeRelayCommand(false, _SelectedDevice.minimumOpenTime));
```

It's never hitting that `var txt = Encoding.ASCII.GetString(data);`  line...

**Reply by aritchie** [reply]
Use the latest preview version.  There was an issue that was fixed with WhenNotificationReceived - https://github.com/shinyorg/shiny/commit/11f3639558975ca58e37ca1747eab245846ccaa5  

Two things, please remove any of your abstracted code like WriteValue when sending anything to the forums.  Second, note that while I'm happy to help, these forums aren't meant for tech support.  Testing BLE without an external device is near impossible, so the user is expected to plugin large gaps here.  The code is open for you to review and things like sourcelink are enabled for debugging.  If you want general support like this, you can go a sponsorship route which will allow you to communicate through a discord channel or I offer consulting in blocks of time.

**Reply by wesoos** [reply]
Thank you for the above.  Everything works great now.  And thank you for all your effort in creating this library.  Really appreciated.

Apologies if I overstepped with this discussion.
